rest/doc_api.go

```diff
@@ -320,6 +330,57 @@ func (h *handler) handlePutDoc() error {
 	return nil
 }
 
+func (h *handler) handlePutDocReplicator2(docid string) (newRev string, doc *db.Document, err error) {
+	bodyBytes, err := h.readBody()
+	if err != nil {
+		return "", nil, err
+	}
+	if bodyBytes == nil || len(bodyBytes) == 0 {
+		return "", nil, base.ErrEmptyDocument
+	}
+
+	newDoc := &db.Document{
+		ID: docid,
+	}
+	newDoc.UpdateBodyBytes(bodyBytes)
+
+	var parentRev string
+	if oldRev := h.getQuery("rev"); oldRev != "" {
+		parentRev = oldRev
+	} else if ifMatch := h.rq.Header.Get("If-Match"); ifMatch != "" {
+		parentRev = ifMatch
+	}
+
+	generation, _ := db.ParseRevID(parentRev)
+	generation++
+
+	newDoc.RevID, _ = db.CreateRevIDWithBytes(generation, parentRev, bodyBytes)
+	history := []string{newDoc.RevID, parentRev}
+	// Handle and pull out expiry
+	if bytes.Contains(bodyBytes, []byte(db.BodyExpiry)) {
+		body := newDoc.Body()
+		expiry, err := body.ExtractExpiry()
+		if err != nil {
+			return "", nil, base.HTTPErrorf(http.StatusBadRequest, "Invalid expiry: %v", err)
+		}
+		newDoc.DocExpiry = expiry
+		newDoc.UpdateBody(body)
+	}
+
```

Comment:
handleRev is setting newDoc.Deleted here based on the message property here - this should do the same based on a query string parameter.